### PR TITLE
vectors

### PR DESCRIPTION
**Pull request by FredTheDino**
Adding in vectors/tuples, most basic operations work for them
and they are a nice way to store data. It's nice!

### CODE REVIEW

File: src/compiler.rs [hunk 1]
```diff
@@ -331,6 +333,51 @@ impl Compiler {
...
 
+    fn grouping_or_tuple(&mut self, block: &mut Block) {
+        let block_length = block.ops.len();
+        let token_length = self.curr;
+        if self.try_tuple(block).is_err() {
+            block.ops.truncate(block_length);
+            self.curr = token_length;
+            self.grouping(block);
+        }
+    }
```
Comment (sornas) [inline src/compiler.rs:344]: Could be a macro seeing as we do this in two places now. `maybe_else!(self.try_truple, self.grouping, block)`

File: src/compiler.rs [hunk 2]
```diff
@@ -331,6 +333,51 @@ impl Compiler {
...
+                    if self.peek() == Token::RightParen {
+                        continue;
+                    }
+                    return Err(());
+                }
+            }
+        }
+        if num_args == 1 {
+            return Err(());
+        }
```
Comment (sornas) [inline src/compiler.rs:374]: What about tuples with one element? Big no or..?

File: src/compiler.rs [hunk 3]
```diff
@@ -331,6 +333,51 @@ impl Compiler {
...
+
+        let mut num_args = 0;
+        loop {
+            match self.peek() {
+                Token::RightParen | Token::EOF => {
+                    break;
+                }
+                Token::Newline => {
+                    self.eat();
+                }
```
Comment (sornas) [inline src/compiler.rs:357]: Wild. Also, we need a language spec.

File: src/compiler.rs [hunk 4]
```diff
@@ -331,6 +333,51 @@ impl Compiler {
...
+                    num_args += 1;
+                    if self.peek() == Token::Comma {
+                        self.eat();
+                        continue;
+                    }
+                    if self.peek() == Token::RightParen {
+                        continue;
+                    }
+                    return Err(());
+                }
```
Comment (sornas) [inline src/compiler.rs:369]:
```suggestion
                token => {
                    self.expression(block);
                    num_args += 1;
                    match token {
                        Token::Comma => { self.eat(); }
                        Token::RightParen => {}
                        _ => { return Err(()); }
                    }
                }
```

File: src/compiler.rs [hunk 5]
```diff
@@ -331,6 +333,51 @@ impl Compiler {
...
+            }
+        }
+        if num_args == 1 {
+            return Err(());
+        }
+
+        expect!(self, Token::RightParen, "Expected ')' after tuple");
```
Comment (sornas) [inline src/compiler.rs:376]:
```suggestion
        expect!(self, Token::RightParen, "Expected ')' after tuple.");
```

File: src/compiler.rs [hunk 6]
```diff
@@ -339,6 +386,15 @@ impl Compiler {
         expect!(self, Token::RightParen, "Expected ')' around expression.");
     }
 
+    fn index(&mut self, block: &mut Block) {
+        expect!(self, Token::LeftBracket, "Expected ']' around index.");
```
Comment (sornas) [inline src/compiler.rs:390]:
```suggestion
        expect!(self, Token::LeftBracket, "Expected '[' around index.");
```

File: src/compiler.rs [hunk 7]
```diff
@@ -339,6 +386,15 @@ impl Compiler {
...
+
+        self.expression(block);
+        block.add(Op::Index, self.line());
+
+        expect!(self, Token::RightBracket, "Expected '[' around index.");
```
Comment (sornas) [inline src/compiler.rs:395]:
```suggestion
        expect!(self, Token::RightBracket, "Expected ']' around index.");
```

File: src/compiler.rs [hunk 8]
```diff
@@ -472,6 +528,9 @@ impl Compiler {
         let mut return_type = Type::Void;
         let mut function_block = Block::new(&name, &self.current_file, self.line());
```
Comment (sornas) [inline src/compiler.rs:529]: This line throws a compiler warning.

File: src/lib.rs [hunk 9]
```diff
@@ -378,10 +387,114 @@ impl Debug for Value {
...
+    }
+
+    pub fn neg(value: &Value) -> Value {
+        match value {
+            Value::Float(a) => Value::Float(-a),
+            Value::Int(a) => Value::Int(-a),
+            Value::Tuple(a) => Value::Tuple(Rc::new(a.iter().map(neg).collect())),
+            _ => Value::Nil,
+        }
+    }
```
Comment (sornas) [inline src/lib.rs:410]: This should be its own PR.

File: src/lib.rs [hunk 10]
```diff
@@ -378,10 +387,114 @@ impl Debug for Value {
...
+
+    pub fn not(value: &Value) -> Value {
+        match value {
+            Value::Bool(a) => Value::Bool(!a),
+            Value::Tuple(a) => Value::Tuple(Rc::new(a.iter().map(not).collect())),
```
Comment (sornas) [inline src/lib.rs:415]: I would love to hear a use case for this. I'm not hard-opposed, but I don't see the value. Or is it just to match that the other operators all work tuples? In any case, I think I prefer Python's `all` and the like (which are easy to expand with more variants) instead of extending the `!`-operator.

File: src/compiler.rs (hunk 2 above)
Comment (FredTheDino) [reply src/compiler.rs:374]: There is no syntax for it but it can be done. Tuples with one element might have uses.

File: src/lib.rs (hunk 9 above)
Comment (FredTheDino) [reply src/lib.rs:410]: I see where you're comming from. But without the operators extracted to functions, the Tuple operations wouldn't work since they're recursive.

File: src/lib.rs (hunk 10 above)
Comment (FredTheDino) [reply src/lib.rs:415]: I just added it for consistency. Since now an operator applied to a tuple, is the same as applying the operator to all it's elements. It's not super usefull since it requires the tuple to be bools, but I don't think this is an alternative to `all`. I see it as a supliment. Note that it returns a tuple.

File: src/compiler.rs (hunk 3 above)
Comment (FredTheDino) [reply src/compiler.rs:357]: Yes.

File: src/compiler.rs (hunk 4 above)
Comment (FredTheDino) [reply src/compiler.rs:369]: This code is not equivalent, but I like the idea.

File: src/compiler.rs (hunk 1 above)
Comment (FredTheDino) [reply src/compiler.rs:344]: I think this is a premature grouping. We might not always want to give `block` to all `try*` functions. There is a nice macro here, but that is a different PR.

File: src/lib.rs (hunk 9 above)
Comment (sornas) [reply src/lib.rs:410]: Yes, but with this it's hard to see the Tuple additions since everything's additions.

File: src/lib.rs (hunk 10 above)
Comment (sornas) [reply src/lib.rs:415]: Oh. That makes more sense actually.

File: src/compiler.rs (hunk 4 above)
Comment (sornas) [reply src/compiler.rs]: It isn't? I tried to make it equivalent.

Comment (sornas) [reply src/compiler.rs]: Oh right I see it now.

File: src/vm.rs [hunk 11]
```diff
@@ -119,12 +119,6 @@ impl VM {
         self.stack.push(value)
     }
 
-    fn pop_twice(&mut self) -> (Value, Value) {
-        let (a, b) = (self.stack.remove(self.stack.len() - 1),
-                      self.stack.remove(self.stack.len() - 1));
-        (b, a)  // this matches the order they were on the stack
-    }
-
```
Comment (sornas) [inline src/vm.rs]: Call this from two_op! instead of making two calls to pop and removing this.

**Review by sornas — APPROVED** [review]
